mastermind.rb

```diff
@@ -1,25 +1,50 @@
 class Mastermind
+  attr_reader :red_count, :white_count
+
   def print_instructions
-    print "Welcome to Mastermind! The object of the game is to guess a 4-color code "\
-        "within 10 guesses. The available colors are red, orange, yellow, green, "\
-        "blue, and purple. Colors may be used in the code more than once or not at all."\
+    puts "\tWelcome to Mastermind! The object of the game is to guess a 4-color code
+        within 10 guesses. The available colors are red, orange, yellow, green,
+        blue, and purple. Colors may be used in the code more than once or not at all.
+
+        To make a guess: type the first letter of each color you're guessing in caps.
+        For example, if you want to guess 'red, orange, yellow, red,' type 'ROYR'.
 
-        "\nTo make a guess: type the first letter of each color you're guessing in caps. "\
-        "For example, if you want to guess 'red, orange, yellow, red,' type 'ROYR'."\
+        After each guess, you will receive feedback in the form of colored pins,
+        letting you know:
+          \t- how many colors are correct but in the wrong position (red pins)
+          \t- how many colors are correct and in the right position (white pins)
 
-        "\nAfter each guess, you will receive feedback in the form of colored pins, letting you know:"\
-          "\n\t- how many colors are correct but in the wrong position (white pins)"\
-          "\n\t- how many colors are correct and in the right position (red pins)"\
+        To quit the game, type '/q'. To start a new game, type '/r'."
+  end
 
-        "\nTo quit the game, type '/q'. To start a new game, type '/r'."
+  def color_count(code)
+    @color_count = Hash.new(0)
+    code.each do |color|
+      color_key = color.to_sym
+      if @color_count.key?(color_key)
+        @color_count[color_key] += 1
+      else
+        @color_count[color_key] = 1
+      end
     end
+    @color_count
+  end
 
-  def compare(code, guess)
+  def get_color_counts(code, guess)
+    code_count = color_count(code)
+    guess = guess.chars
+    guess_count = color_count(guess)
+  end
+
+  def compare_red_pins(code_color_count, guess_color_count)
+    @red_count = 0
+    guess_count.each do |key, value|
+    end
   end
 end
 
-class Code
-  attr_reader :code
+class SecretCode
+  attr_reader :code, :color_count
```

Comment:
Thanks, Pat! I like the elegance of `map` (it works in Ruby just like it does in Python!) and appreciate the explanation of `|_|`. Is it the case that you'd have an actual variable name if the function we're mapping to is more complex? 

And thank you for introducing me to `sample`!

Reply:
Would it be better to have my `code` method as an `initialize` method? Seemingly we'll always need a code generated and perhaps it's more efficient to get a new code by calling `c = SecretCode.new` as opposed to `c = SecretCode.new.code`.

Reply:
Yeah, you would just substitute the `_` with a variable name if you wanted to use it. For example:

```rb
[1, 2, 3].map { |num| num + 1 }
# [2, 3, 4]
```

Reply:
Since you're using a class, I think you're only option is to have some sort of `code` method/variable that can be accessed to provide the code. You _could_ set `code` as some sort of variable in your initialize method, but I think it would be cleaner to just have a `code` method.